### ISSUE TITLE
Problem with predictions in txt files using CLI

### ISSUE THREAD

**Issue by philosGit** [ultralytics/ultralytics#6209]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

Hello,

I am using the CLI (according to https://docs.ultralytics.com/usage/cli/) to train a YOLOv8-seg model for multiple classes on custom data. The training is working just fine and now I wanted to use the trained model to make predictions for my data. The prediction results look good to me (based on visual inspection from output images). I need the polygon points of the predictions of each individual class. Therefore, I set the argument save_txt to true for the prediction with the CLI as described in the documentation. The text files are created and filled, but now the problem is that the polygon points are not correct in most of my cases (when one class is part of another). For this reason I created a short example image (yolov8-seg_cli_prediction_example.pdf) to give a better idea of the problem. I am sorry about not sharing the actual data I am using, but I am not allowed too. I hope you can unterstand this and are still able to get an idea from the example image.

Example image:
The example image shows three different cases with three example classes. Hereby the class to point out the problem is class 1, while the object from class 3 is causing the problem. Cases:
1.) The detected object of the 3rd class is fully enclosed within the 1st class object and is not touching the polygon border. In this case the text file output is as expected (labeled like this during training).
2.) The detected object of the 3rd class lies within the 1st class object, but shares the same border on one side (could also be slightly extended). In this case the returned polygon points for the 1st class in the text file include the 3rd class object as a cut-out.
3.) The detected object of the 3rd class lies within the 1st class object, but shares the same border on two sides (could also be slightly extended). In this case only one part of the divided 1st class polygon is returned as shown.

For me also the second case would be fine (if other polygon gets cut out), but I would still need at least the other part of the polygon points in third case saved to the text file. I would really appreciate your help, since using the CLI is very easy and saves time. Else I would try a workaround (train multiple models, etc.). Thank you very much in advance.

[yolov8-seg_cli_prediction_example.pdf](https://github.com/ultralytics/ultralytics/files/13298744/yolov8-seg_cli_prediction_example.pdf)

### Environment

OS                  Windows-10-10.0.22621-SP0
Environment         Windows
Python              3.11.5
Install             pip
RAM                 15.24 GB
CPU                 12th Gen Intel Core(TM) i7-1260P
CUDA                None

matplotlib           3.8.0>=3.3.0
numpy                1.26.0>=1.22.2
opencv-python        4.8.0.76>=4.6.0
pillow               10.0.1>=7.1.2
pyyaml               6.0.1>=5.3.1
requests             2.31.0>=2.23.0
scipy                1.11.2>=1.4.1
torch                2.0.1>=1.8.0
torchvision          0.15.2>=0.9.0
tqdm                 4.66.1>=4.64.0
pandas               2.1.1>=1.1.4
seaborn              0.12.2>=0.11.0
psutil               5.9.5
py-cpuinfo           9.0.0

### Minimal Reproducible Example

yolo predict model=runs/segment/yolov8m-seg/weights/best.pt source=datasets/test/images save_txt=True

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@philosGit hello,

Thank you for providing a detailed description of the issue you're encountering with polygon predictions using YOLOv8 segmentation in the CLI. Understanding the nuances of your problem is crucial, especially since you're dealing with instances where one class is part of another, which can certainly complicate the segmentation predictions.

Based on your example image and explanation, it seems that the primary issue arises when overlapping or adjoining boundaries are present between different classes. The first scenario working as expected is certainly encouraging, and it suggests that the model's ability to discern distinct objects isn't a fundamental limitation but rather a question of how the boundaries are handled during prediction.

The challenge you've described in cases 2 and 3, particularly with the second object cutting out from the first or missing polygon parts, is indicative of how the post-processing is handling intersecting classes. This is an area where the internal logic for polygon segmentation may need to be reviewed and potentially adjusted to ensure that the model faithfully represents the objects as trained, even in cases of adjoining boundaries.

Considering you're using the CLI—which is designed to be straightforward and time-efficient—it is understandable that diving into code alterations might not be your preferred approach. However, since you've already trained a model that visually seems to produce satisfactory results, you might want to consider exploring the post-processing environment for better control over how the model resolves these overlapping boundaries.

Although you’re not able to share the actual data, your explanations and example cases offer a reasonable starting point for investigating this behavior further. It would be beneficial to consider if there are any patterns in how these overlapping cases are resolved and if there's a consistent logic that might be exploited to correct the output.

Given the significant impact that this issue has on the usability of the output for your use case, it’s worth raising this as a potential area for enhancement in future updates. Engaging with developers and other users who might have faced similar issues could lead to a collective effort in refining this aspect of YOLOv8's segmentation capability.

Thank you once again for reporting this. Although we cannot directly resolve this in the short term without eliciting further discussion and potentially adjustments in the model's post-processing logic, your constructive input is immensely valuable for continual improvement of YOLOv8. We appreciate your understanding and cooperation.

Best regards,
The YOLOv8 Development Team

**Comment by philosGit** [comment]
@glenn-jocher hello,

Thank you very much for your fast reply and your effort. It is understandable that some time is needed to investigate this and hopefully fix it in the future. I will have a look in alternative approaches as mentioned by you (post-processing environment, train one model per class, etc.). 

Should I close the issue or will it be kept open until it is resolved?

Thanks again and best regards!

**Comment by glenn-jocher** [comment]
@philosGit hello,

Thank you for your understanding and patience as we delve into the challenges you're facing with YOLOv8 segmentation. I'm glad to hear that you're considering alternative approaches in the meantime; it's always a good strategy to explore multiple avenues to achieve the best possible outcomes for your use case.

Regarding the GitHub issue, I believe it would be beneficial to keep it open for now. This will allow us and the community to track the issue more effectively, and it may also invite input from other users who might be encountering similar challenges or who might have found their own workarounds. Keeping the issue open also serves as a reminder for our team to revisit this when we have updates or potential fixes.

We sincerely appreciate your active participation and contributions to improving YOLOv8. If you do experiment with different strategies or make observations that could be helpful, please feel free to update the issue with your findings. Collective insights from our user community are invaluable in making YOLOv8 better for everyone.

Once again, thank you for your collaboration and best wishes as you continue working on your project.

Best regards,
The YOLOv8 Development Team

**Comment by philosGit** [comment]
**My workaround:**

I would still like to use the CLI, since it is straightforward, time-efficient and it is easy to use by other people. Therefore, I investigated the possible arguments, which can be used for the prediction (https://docs.ultralytics.com/modes/predict/#inference-arguments) and saw that there is an argument 'classes'. I used this argument also for the training to train a model for each of my three classes. Hereby note that the argument classes is currently missing from the table (https://docs.ultralytics.com/modes/train/#arguments), but it still works as expected. For me this approach worked well, since I don't have lots of data and only three classes, but of course the training time needed to train multiple models is the limiting factor here. Still, I hope this helps people, who also need an easy workaround to this problem.

**Comment by glenn-jocher** [comment]
Hello @philosGit,

Thank you for sharing your workaround with the community! It's interesting to hear that you've sought to utilize the 'classes' argument for prediction to compartmentalize the training for each of your classes independently. This is indeed a creative application that can be particularly effective in scenarios where the dataset is manageable and there are a few distinct classes.

Your point about the missing 'classes' argument in the training documentation is well-taken, and we apologize for any inconvenience this oversight may have caused. We will be sure to review our documentation and ensure consistency across our resources, so all users can benefit from complete and accurate information.

While training separate models for each class can introduce longer training times, as you’ve noted, it is certainly an approach that can provide a practical solution in the short term, especially when simplicity and ease of use are high priorities.

We appreciate your input, and it's great to see members of the community proactively finding ways to push through their challenges. Your contribution may indeed help others facing similar issues. If you have any further insights or feedback as you continue to work with YOLOv8, please don't hesitate to share.

We're committed to improving YOLOv8, and user experiences like yours are invaluable to our ongoing development process. Thank you for being part of our community, and we wish you success with your continued use of YOLOv8.

Best wishes,
The YOLOv8 Team

**Comment by glenn-jocher** [comment]
Hello @philosGit,

That's fantastic investigative work on your end! Pinpointing the `overlap_mask` setting as the source of the segmentation issue is a significant breakthrough and offers a much more streamlined solution to your problem. Your effort in diving into the documentation and understanding the implications of different training arguments exemplifies the merit and power of a well-informed user community.

Disabling `overlap_mask` indeed changes how overlapping classes are handled during the segmentation mask generation process, which appears to align better with your specific needs. This solution has the advantage of maintaining a single model while producing the desired output, which is efficient and simplifies the inference process.

Your update is sure to assist others within the Ultralytics community who may face similar challenges. Understanding the role and impact of each training argument helps to fine-tune models effectively, leading to more accurate and desirable outcomes as you have achieved.

We'll note your solution for reference in our development discussions and documentation updates. This kind of feedback is invaluable in helping us refine our models and enhance user experience.

Feel free to mark the issue as resolved, and thank you for your contributions to making YOLOv8 better for everyone. We're glad you're part of the Ultralytics community, and we wish you all the best with your ongoing projects!

Enjoy your time developing as well!
